lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp

```diff
@@ -139,285 +135,379 @@
+  assert(L->contains(exiting_block));
+  assert(br->isConditional());
+  BasicBlock *result = L->contains(br->getSuccessor(0)) ? br->getSuccessor(1) : br->getSuccessor(0);
+  assert(!L->contains(result));
+  return result;
+}
+
+// Branch over the block's content with the condition cond.
+// All values used outside the block is replaced by a phi.
+//
+static void SkipBlockWithBranch(BasicBlock *bb, Value *cond, Loop *L, LoopInfo *LI) {
+  BasicBlock *body = bb->splitBasicBlock(bb->getFirstNonPHI());
+  body->setName("dx.struct_exit.cond_body");
+  BasicBlock *end = body->splitBasicBlock(body->getTerminator());
+  end->setName("dx.struct_exit.cond_end");
+
+  bb->getTerminator()->eraseFromParent();
+  BranchInst::Create(end, body, cond, bb);
+
+  for (Instruction &inst : *body) {
+    PHINode *phi = nullptr;
+
+    for (User *user : inst.users()) {
+      Instruction *user_inst = dyn_cast<Instruction>(user);
+      if (!user_inst)
+        continue;
+
+      if (user_inst->getParent() != body) {
+        if (!phi) {
+          phi = PHINode::Create(inst.getType(), 2, "", &*end->begin());
+          phi->addIncoming(UndefValue::get(inst.getType()), bb);
+          phi->addIncoming(&inst, body);
+        }
+
+        user_inst->replaceUsesOfWith(&inst, phi);
+      }
+    } // For each user of inst of body
+  } // For each inst in body
+
+  L->addBasicBlockToLoop(body, *LI);
+  L->addBasicBlockToLoop(end, *LI);
+}
+
+static unsigned GetNumPredecessors(BasicBlock *bb) {
+  unsigned ret = 0;
+  for (BasicBlock *pred : llvm::predecessors(bb)) {
+    (void)pred;
+    ret++;
   }
+  return ret;
+}
 
-  // If bb has side effect, split it into 3 basic blocks, where its body is
-  // gated behind if (!exit_cond)
-  for (std::pair<BasicBlock *, Value *> data : blocks_with_side_effect) {
-    BasicBlock *bb = data.first;
-    Value *exit_cond = data.second;
+static bool RemoveUnstructuredLoopExitsIteration(BasicBlock *exiting_block, Loop *L, LoopInfo *LI, DominatorTree *DT) {
 
-    BasicBlock *body = bb->splitBasicBlock(bb->getFirstNonPHI());
-    body->setName("dx.struct_exit.cond_body");
-    BasicBlock *end = body->splitBasicBlock(body->getTerminator());
-    end->setName("dx.struct_exit.cond_end");
+  LLVMContext &ctx = L->getHeader()->getContext();
+  Type *i1Ty = Type::getInt1Ty(ctx);
 
-    bb->getTerminator()->eraseFromParent();
-    BranchInst::Create(end, body, exit_cond, bb);
+  BasicBlock *exit_block = GetExitBlockForExitingBlock(L, exiting_block);
 
-    for (Instruction &inst : *body) {
-      PHINode *phi = nullptr;
+  BasicBlock *latch = L->getLoopLatch();
+  BasicBlock *latch_exit = GetExitBlockForExitingBlock(L, latch);
 
-      for (User *user : inst.users()) {
-        Instruction *user_inst = dyn_cast<Instruction>(user);
-        if (!user_inst)
-          continue;
+  // If exiting block already dominates latch, then no need to do anything.
+  if (exiting_block == latch || DT->dominates(exiting_block, latch)) {
```

Comment:
The call to `dominates` is enough, it's `properlyDominates` which would return `false` if the blocks are the same.